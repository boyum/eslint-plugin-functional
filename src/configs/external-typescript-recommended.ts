import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import externalVanillaRecommended from "#eslint-plugin-functional/configs/external-vanilla-recommended";

const tsConfig = {
  "@typescript-eslint/prefer-readonly": "error",
  "@typescript-eslint/switch-exhaustiveness-check": "error",
} satisfies FlatConfig.Config["rules"];

export default {
  ...externalVanillaRecommended,
  ...tsConfig,
} satisfies FlatConfig.Config["rules"];
