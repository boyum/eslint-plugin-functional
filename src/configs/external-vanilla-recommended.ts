import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

export default {
  "prefer-const": "error",
  "no-param-reassign": "error",
  "no-var": "error",
} satisfies FlatConfig.Config["rules"];
