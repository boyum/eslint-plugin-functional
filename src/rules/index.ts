import * as functionalParameters from "./functional-parameters";
import * as immutableData from "./immutable-data";
import * as noClassInheritance from "./no-class-inheritance";
import * as noClasses from "./no-classes";
import * as noConditionalStatements from "./no-conditional-statements";
import * as noExpressionStatements from "./no-expression-statements";
import * as noLet from "./no-let";
import * as noLoopStatements from "./no-loop-statements";
import * as noMixedTypes from "./no-mixed-types";
import * as noPromiseReject from "./no-promise-reject";
import * as noReturnVoid from "./no-return-void";
import * as noThisExpressions from "./no-this-expressions";
import * as noThrowStatements from "./no-throw-statements";
import * as noTryStatements from "./no-try-statements";
import * as preferImmutableTypes from "./prefer-immutable-types";
import * as preferPropertySignatures from "./prefer-property-signatures";
import * as preferReadonlyTypes from "./prefer-readonly-type";
import * as preferTacit from "./prefer-tacit";
import * as readonlyType from "./readonly-type";
import * as typeDeclarationImmutability from "./type-declaration-immutability";

/**
 * All of the custom rules.
 */
export const rules: Readonly<{
  [functionalParameters.name]: typeof functionalParameters.rule;
  [immutableData.name]: typeof immutableData.rule;
  [noClasses.name]: typeof noClasses.rule;
  [noClassInheritance.name]: typeof noClassInheritance.rule;
  [noConditionalStatements.name]: typeof noConditionalStatements.rule;
  [noExpressionStatements.name]: typeof noExpressionStatements.rule;
  [noLet.name]: typeof noLet.rule;
  [noLoopStatements.name]: typeof noLoopStatements.rule;
  [noMixedTypes.name]: typeof noMixedTypes.rule;
  [noPromiseReject.name]: typeof noPromiseReject.rule;
  [noReturnVoid.name]: typeof noReturnVoid.rule;
  [noThisExpressions.name]: typeof noThisExpressions.rule;
  [noThrowStatements.name]: typeof noThrowStatements.rule;
  [noTryStatements.name]: typeof noTryStatements.rule;
  [preferImmutableTypes.name]: typeof preferImmutableTypes.rule;
  [preferPropertySignatures.name]: typeof preferPropertySignatures.rule;
  [preferReadonlyTypes.name]: typeof preferReadonlyTypes.rule;
  [preferTacit.name]: typeof preferTacit.rule;
  [readonlyType.name]: typeof readonlyType.rule;
  [typeDeclarationImmutability.name]: typeof typeDeclarationImmutability.rule;
}> = {
  [functionalParameters.name]: functionalParameters.rule,
  [immutableData.name]: immutableData.rule,
  [noClasses.name]: noClasses.rule,
  [noClassInheritance.name]: noClassInheritance.rule,
  [noConditionalStatements.name]: noConditionalStatements.rule,
  [noExpressionStatements.name]: noExpressionStatements.rule,
  [noLet.name]: noLet.rule,
  [noLoopStatements.name]: noLoopStatements.rule,
  [noMixedTypes.name]: noMixedTypes.rule,
  [noPromiseReject.name]: noPromiseReject.rule,
  [noReturnVoid.name]: noReturnVoid.rule,
  [noThisExpressions.name]: noThisExpressions.rule,
  [noThrowStatements.name]: noThrowStatements.rule,
  [noTryStatements.name]: noTryStatements.rule,
  [preferImmutableTypes.name]: preferImmutableTypes.rule,
  [preferPropertySignatures.name]: preferPropertySignatures.rule,
  [preferReadonlyTypes.name]: preferReadonlyTypes.rule,
  [preferTacit.name]: preferTacit.rule,
  [readonlyType.name]: readonlyType.rule,
  [typeDeclarationImmutability.name]: typeDeclarationImmutability.rule,
};
