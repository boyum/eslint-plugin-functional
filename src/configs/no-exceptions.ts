import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#eslint-plugin-functional/rules";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

export default Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.category === "No Exceptions" &&
        rule.meta.docs.recommended !== false,
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];
