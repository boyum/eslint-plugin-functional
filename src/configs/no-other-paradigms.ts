import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#eslint-plugin-functional/rules";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

export default Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.category === "No Other Paradigms" &&
        rule.meta.docs.recommended !== false,
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];
