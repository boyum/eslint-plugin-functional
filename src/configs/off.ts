import type { FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import { rules } from "#/rules";
import { ruleNameScope } from "#/utils/misc";

export default Object.fromEntries(
  Object.entries(rules).map(([name]) => [`${ruleNameScope}/${name}`, "off"]),
) satisfies FlatConfig.Config["rules"] as NonNullable<FlatConfig.Config["rules"]>;
