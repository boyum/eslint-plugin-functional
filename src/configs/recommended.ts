import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";
import { Immutability } from "is-immutable-type";

import { rules } from "#eslint-plugin-functional/rules";
import * as functionalParameters from "#eslint-plugin-functional/rules/functional-parameters";
import * as noConditionalStatements from "#eslint-plugin-functional/rules/no-conditional-statements";
import * as noLet from "#eslint-plugin-functional/rules/no-let";
import * as noThisExpressions from "#eslint-plugin-functional/rules/no-this-expressions";
import * as noThrowStatements from "#eslint-plugin-functional/rules/no-throw-statements";
import * as noTryStatements from "#eslint-plugin-functional/rules/no-try-statements";
import * as preferImmutableTypes from "#eslint-plugin-functional/rules/prefer-immutable-types";
import * as typeDeclarationImmutability from "#eslint-plugin-functional/rules/type-declaration-immutability";
import { RuleEnforcementComparator } from "#eslint-plugin-functional/rules/type-declaration-immutability";
import { ruleNameScope } from "#eslint-plugin-functional/utils/misc";

const recommended = Object.fromEntries(
  Object.entries(rules)
    .filter(
      ([, rule]) =>
        rule.meta.deprecated !== true &&
        rule.meta.docs.recommended === "recommended" &&
        rule.meta.docs.category !== "Stylistic",
    )
    .map(([name, rule]) => [
      `${ruleNameScope}/${name}`,
      rule.meta.docs.recommendedSeverity,
    ]),
) satisfies FlatConfig.Config["rules"];

const overrides = {
  [functionalParameters.fullName]: [
    "error",
    {
      enforceParameterCount: {
        ignoreLambdaExpression: true,
        ignoreIIFE: true,
        ignoreGettersAndSetters: true,
      },
    },
  ],
  [noConditionalStatements.fullName]: [
    "error",
    {
      allowReturningBranches: true,
    },
  ],
  [noLet.fullName]: [
    "error",
    {
      allowInForLoopInit: true,
    },
  ],
  [noThisExpressions.fullName]: "off",
  [noThrowStatements.fullName]: [
    "error",
    {
      allowInAsyncFunctions: true,
    },
  ],
  [noTryStatements.fullName]: "off",
  [preferImmutableTypes.fullName]: [
    "error",
    {
      enforcement: "None",
      ignoreInferredTypes: true,
      parameters: {
        enforcement: "ReadonlyDeep",
      },
    },
  ],
  [typeDeclarationImmutability.fullName]: [
    "error",
    {
      rules: [
        {
          identifiers: ["^I?Immutable.+"],
          immutability: Immutability.Immutable,
          comparator: RuleEnforcementComparator.AtLeast,
        },
        {
          identifiers: ["^I?ReadonlyDeep.+"],
          immutability: Immutability.ReadonlyDeep,
          comparator: RuleEnforcementComparator.AtLeast,
        },
        {
          identifiers: ["^I?Readonly.+"],
          immutability: Immutability.ReadonlyShallow,
          comparator: RuleEnforcementComparator.AtLeast,
          fixer: [
            {
              pattern: "^(Array|Map|Set)<(.+)>$",
              replace: "Readonly$1<$2>",
            },
            {
              pattern: "^(.+)$",
              replace: "Readonly<$1>",
            },
          ],
        },
        {
          identifiers: ["^I?Mutable.+"],
          immutability: Immutability.Mutable,
          comparator: RuleEnforcementComparator.AtMost,
          fixer: [
            {
              pattern: "^Readonly(Array|Map|Set)<(.+)>$",
              replace: "$1<$2>",
            },
            {
              pattern: "^Readonly<(.+)>$",
              replace: "$1",
            },
          ],
        },
      ],
    },
  ],
} satisfies FlatConfig.Config["rules"];

export default {
  ...recommended,
  ...overrides,
};
