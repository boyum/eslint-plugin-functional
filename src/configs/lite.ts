import { type FlatConfig } from "@typescript-eslint/utils/ts-eslint";

import * as functionalParameters from "#eslint-plugin-functional/rules/functional-parameters";
import * as immutableData from "#eslint-plugin-functional/rules/immutable-data";
import * as noConditionalStatements from "#eslint-plugin-functional/rules/no-conditional-statements";
import * as noExpressionStatements from "#eslint-plugin-functional/rules/no-expression-statements";
import * as preferImmutableTypes from "#eslint-plugin-functional/rules/prefer-immutable-types";

import recommended from "./recommended";

const overrides = {
  [functionalParameters.fullName]: [
    "error",
    {
      enforceParameterCount: false,
    },
  ],
  [immutableData.fullName]: ["error", { ignoreClasses: "fieldsOnly" }],
  [noConditionalStatements.fullName]: "off",
  [noExpressionStatements.fullName]: "off",
  [preferImmutableTypes.fullName]: [
    "error",
    {
      enforcement: "None",
      ignoreInferredTypes: true,
      parameters: {
        enforcement: "ReadonlyShallow",
      },
    },
  ],
} satisfies FlatConfig.Config["rules"];

export default {
  ...recommended,
  ...overrides,
};
