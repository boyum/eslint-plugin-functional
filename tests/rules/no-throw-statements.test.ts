import dedent from "dedent";
import { createRuleTester } from "eslint-vitest-rule-tester";
import { describe, expect, it } from "vitest";

import { name, rule } from "#/rules/no-throw-statements";

import { esLatestConfig } from "../utils/configs";

describe(name, () => {
  describe("javascript - es latest", () => {
    const { valid, invalid } = createRuleTester({
      name,
      rule,
      configs: esLatestConfig,
    });

    it("doesn't report non-issues", () => {
      valid(dedent`
        function foo() {
          bar();
        }
      `);
    });

    it("reports throw statements of strings", () => {
      const invalidResult = invalid({
        code: dedent`
          function foo() {
            throw 'error';
          }
        `,
        errors: ["generic"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports throw statements of Errors", () => {
      const invalidResult = invalid({
        code: dedent`
          function foo() {
            throw new Error();
          }
        `,
        errors: ["generic"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });

    it("reports throw statements in async functions", () => {
      const invalidResult = invalid({
        code: dedent`
          async function foo() {
            throw new Error();
          }
        `,
        errors: ["generic"],
      });
      expect(invalidResult.messages).toMatchSnapshot();
    });
  });
});
