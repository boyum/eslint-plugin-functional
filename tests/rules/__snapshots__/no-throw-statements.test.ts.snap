// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`no-throw-statements > javascript - es latest > options > allowToRejectPromises > reports throw statements in functions nested in async functions 1`] = `
[
  {
    "column": 5,
    "endColumn": 23,
    "endLine": 3,
    "line": 3,
    "message": "Unexpected throw, throwing exceptions is not functional.",
    "messageId": "generic",
    "nodeType": "ThrowStatement",
    "ruleId": "no-throw-statements",
    "severity": 2,
  },
]
`;

exports[`no-throw-statements > javascript - es latest > options > allowToRejectPromises > reports throw statements in try with catch in async functions 1`] = `
[
  {
    "column": 5,
    "endColumn": 36,
    "endLine": 3,
    "line": 3,
    "message": "Unexpected throw, throwing exceptions is not functional.",
    "messageId": "generic",
    "nodeType": "ThrowStatement",
    "ruleId": "no-throw-statements",
    "severity": 2,
  },
]
`;

exports[`no-throw-statements > javascript - es latest > reports throw statements in async functions 1`] = `
[
  {
    "column": 3,
    "endColumn": 21,
    "endLine": 2,
    "line": 2,
    "message": "Unexpected throw, throwing exceptions is not functional.",
    "messageId": "generic",
    "nodeType": "ThrowStatement",
    "ruleId": "no-throw-statements",
    "severity": 2,
  },
]
`;

exports[`no-throw-statements > javascript - es latest > reports throw statements of Errors 1`] = `
[
  {
    "column": 3,
    "endColumn": 21,
    "endLine": 2,
    "line": 2,
    "message": "Unexpected throw, throwing exceptions is not functional.",
    "messageId": "generic",
    "nodeType": "ThrowStatement",
    "ruleId": "no-throw-statements",
    "severity": 2,
  },
]
`;

exports[`no-throw-statements > javascript - es latest > reports throw statements of strings 1`] = `
[
  {
    "column": 3,
    "endColumn": 17,
    "endLine": 2,
    "line": 2,
    "message": "Unexpected throw, throwing exceptions is not functional.",
    "messageId": "generic",
    "nodeType": "ThrowStatement",
    "ruleId": "no-throw-statements",
    "severity": 2,
  },
]
`;
